rfd/0167-auto-updates-change-proposal.md

```diff
@@ -0,0 +1,179 @@
+
+Issue: Some users will attempt to install the latest available teleport-ent package from the stable/cloud repository. This results in a failed installation if the Teleport client is incompatible with the Teleport control plane.
+
+Issue: Some users will attempt to manually update the teleport-ent package to the latest available version in the stable/cloud repository. This results in a failed update if the Teleport client is incompatible with the Teleport control plane.
+
+### Helm Charts
+The teleport/teleport-kube-agent Helm release manages the deployment of both the teleport-agent and the teleport-agent-updater. This may be an issue for some users become the teleport-agent-updater updates the image of the teleport-agent. This results in the version of the teleport-agent diverging from the version specific in the Helm chart.
+
+Issue: There are a number of users unable to enroll in automatic updates because it is incompatible with their ArgoCD deployments. ArgoCD generates the template from the Helm chart and manages the resources itself. It monitors the teleport-agent resource and when it detects that the resource has diverged from the initial spec, it will reconcile the resource.
+
+### Test Coverage
+Because it was assumed that the Teleport updater would be stable, the Teleport updater logic is written in bash and it lacks sufficient testing. There is some automated testing. There are unit tests in place to verify the functionality of the Teleport updater, and there are tests to verify that the Teleport updater can be installed from the stable/cloud repository.
+
+Issue: For such a critical piece of the Teleport architecture, it seems like an insufficient amount of test coverage. Developers are uncomfortable making changes to the Teleport updater.
+
+### Self-Hosted Support
+The automatic updates feature has been pushed for Teleport Cloud clusters, but not for self-hosted clusters. This has lead development of special execution paths in the Teleport codebase just for Teleport Cloud. These paths are typically not test as extensively, and are expensive to maintain.
+
+Issue: This has lead to some incidents where changes made specifically for Teleport Cloud were merged but contained a bug in a specific edge case. The edge case was not thoroughly tested and the bug was deployed.
+
+### Integrations
+The automatic updates feature does not support Teleport integrations. The Teleport Operator was introduced in Teleport 15 and is already being used for self-hosted plugins. As adoption increases, users may face the same issues with keeping their integrations updated.
+
+## Change proposals
```

Comment:
Things to add:
- Upgrader should not have its own timer. As soon as the proxy serves a new agent version, the agent is installed.
- The agent version served by the proxy should be configurable. I would suggest proxy read the version from a monitored file on disk, so that it's easy to update via Kubernetes / Cloud. tctl would work as well, but it's harder to change via Kubernetes.